File: leetcode/queue_stack/3476_Maximize_Profit_from_Task_Assignment/codes/arr.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    long long maxProfit(vector<int>& workers, vector<vector<int>>& tasks) {
        unordered_map<int, priority_queue<int>> mp;
        for (auto &t : tasks) {
            mp[t[0]].push(t[1]);
        }
        long long total = 0;
        for (int w : workers) {
            if (mp.count(w) && !mp[w].empty()) {
                total += mp[w].top();
                mp[w].pop();
            }
        }
        int maxExtra = 0;
        for (auto &kv : mp) {
            if (!kv.second.empty()) {
                maxExtra = max(maxExtra, kv.second.top());
            }
        }
        total += maxExtra;
        return total;
    }
};

File: leetcode/queue_stack/3476_Maximize_Profit_from_Task_Assignment/codes/arr.py
from typing import List
from collections import defaultdict

class Solution:
    def maxProfit(self, workers: List[int], tasks: List[List[int]]) -> int:
        skill_to_profits = defaultdict(list)
        for s, p in tasks:
            skill_to_profits[s].append(p)
        for s in skill_to_profits:
            skill_to_profits[s].sort(reverse=True)
        used = set()
        total = 0
        for w in workers:
            if w in skill_to_profits and skill_to_profits[w]:
                total += skill_to_profits[w][0]
                used.add((w, 0))
                skill_to_profits[w].pop(0)
        max_extra = 0
        for s in skill_to_profits:
            if skill_to_profits[s]:
                max_extra = max(max_extra, skill_to_profits[s][0])
        total += max_extra
        return total
